File: src/main/java/com/freya02/docs/FieldDoc.java
package com.freya02.docs;

import com.freya02.bot.utils.HTMLElement;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.jsoup.nodes.Element;

public class FieldDoc {
	@NotNull private final ClassDoc classDocs;

	@NotNull private final String fieldName;
	@NotNull private final String fieldType;
	@Nullable private final HTMLElement descriptionElement;

	@NotNull private final String elementId;
	@NotNull private final String url;
	@NotNull private final String modifiers;

	@NotNull private final DetailToElementsMap detailToElementsMap;

	@Nullable private final HTMLElement incubatingElement;
	@Nullable private final SeeAlso seeAlso;

	public FieldDoc(@NotNull ClassDoc classDocs, @NotNull Element element) {
		this.classDocs = classDocs;

		this.elementId = element.id();
		this.url = classDocs.getURL() + "#" + elementId;

		//Get field modifiers
		final Element modifiersElement = element.selectFirst("div.member-signature > span.modifiers");
		if (modifiersElement == null) throw new DocParseException();
		this.modifiers = modifiersElement.text();

		//Get field name
		final Element fieldNameElement = element.selectFirst("h3");
		if (fieldNameElement == null) throw new DocParseException();
		this.fieldName = fieldNameElement.text();

		//Get field type
		final Element fieldTypeElement = element.selectFirst("div.member-signature > span.return-type");
		if (fieldTypeElement == null) throw new DocParseException();
		this.fieldType = fieldTypeElement.text();

		//Get field description
		final Element descriptionElement = element.selectFirst("section.detail > div.block");
		if (descriptionElement != null) {
			this.descriptionElement = new HTMLElement(descriptionElement);
		} else {
			this.descriptionElement = null;
		}

		this.detailToElementsMap = new DetailToElementsMap(element);

		this.incubatingElement = detailToElementsMap.findFirst(DocDetailType.INCUBATING);

		final HTMLElement seeAlsoElement = detailToElementsMap.findFirst(DocDetailType.SEE_ALSO);
		if (seeAlsoElement != null) {
			this.seeAlso = new SeeAlso(seeAlsoElement);
		} else {
			this.seeAlso = null;
		}

		detailToElementsMap.onParseFinished();
	}

	@NotNull
	public ClassDoc getClassDocs() {
		return classDocs;
	}

	@NotNull
	public String getElementId() {
		return elementId;
	}

	@NotNull
	public String getModifiers() {
		return modifiers;
	}

	@NotNull
	public String getFieldName() {
		return fieldName;
	}

	@NotNull
	public String getFieldType() {
		return fieldType;
	}

	@Nullable
	public HTMLElement getDescriptionElement() {
		return descriptionElement;
	}

	@Override
	public String toString() {
		return fieldType + " " + fieldName + " : " + descriptionElement;
	}

	@NotNull
	public String getURL() {
		return url;
	}

	public String getSimpleSignature() {
		return fieldType + " " + fieldName;
	}

	public DetailToElementsMap getDetailToElementsMap() {
		return detailToElementsMap;
	}

	@Nullable
	public SeeAlso getSeeAlso() {
		return seeAlso;
	}
}

File: doxxy-docs/src/main/kotlin/dev/freya02/doxxy/docs/data/FieldDoc.kt
package dev.freya02.doxxy.docs.data

import dev.freya02.doxxy.docs.ClassDocs
import dev.freya02.doxxy.docs.DocParseException
import dev.freya02.doxxy.docs.HTMLElement
import dev.freya02.doxxy.docs.HTMLElementList
import io.github.oshai.kotlinlogging.KotlinLogging
import org.jsoup.nodes.Element

private val logger = KotlinLogging.logger { }

class FieldDoc(val classDocs: ClassDoc, val classDetailType: ClassDetailType, element: Element) : BaseDoc() {
    override val effectiveURL: String
    override val onlineURL: String?

    val fieldName: String
    val fieldType: String
    val fieldValue: String?

    override val descriptionElements: HTMLElementList
    override val deprecationElement: HTMLElement?

    val elementId: String
    val modifiers: String

    override val detailToElementsMap: DetailToElementsMap

    override val seeAlso: SeeAlso?

    init {
        elementId = element.id()
        effectiveURL = classDocs.effectiveURL + "#" + elementId
        onlineURL = classDocs.onlineURL?.let { "$it#$elementId" }

        //Get field modifiers
        val modifiersElement = element.selectFirst("div.member-signature > span.modifiers") ?: throw DocParseException()
        modifiers = modifiersElement.text()

        //Get field name
        val fieldNameElement = element.selectFirst("h3") ?: throw DocParseException()
        fieldName = fieldNameElement.text()

        //Get field type
        val fieldTypeElement =
            element.selectFirst("div.member-signature > span.return-type") ?: throw DocParseException()
        fieldType = fieldTypeElement.text()

        //Get field description
        descriptionElements = HTMLElementList.fromElements(element.select("section.detail > div.block"))

        //Get field possible's deprecation
        deprecationElement = HTMLElement.tryWrap(element.selectFirst("section.detail > div.deprecation-block"))

        //Details
        detailToElementsMap = DetailToElementsMap.parseDetails(element)

        //See also
        val seeAlsoDetail = detailToElementsMap.getDetail(DocDetailType.SEE_ALSO)
        seeAlso = when {
            seeAlsoDetail != null -> SeeAlso(classDocs.source, seeAlsoDetail)
            else -> null
        }

        fieldValue = when {
            "static" in modifiers //public might be omitted in interface constants
                    && "final" in modifiers
                    && seeAlso != null
                    && seeAlso.getReferences().any { it.text == "Constant Field Values" && it.link.contains("/constant-values.html") } -> {
                val constantsMap = ClassDocs.getSource(classDocs.source).getFqcnToConstantsMap()[classDocs.classNameFqcn]
                when {
                    constantsMap != null -> constantsMap[fieldName]
                    else -> {
                        logger.warn { "Could not find constants in ${classDocs.classNameFqcn}" }
                        null
                    }
                }
            }
            else -> null
        }
    }

    override fun toString(): String {
        return "$fieldType $fieldName : $descriptionElements"
    }

    val simpleSignature: String
        get() = fieldName

    override val className: String
        get() = classDocs.className

    override val identifier: String
        get() = simpleSignature
    override val identifierNoArgs: String
        get() = simpleSignature
    override val humanIdentifier: String
        get() = simpleSignature

    override val returnType: String
        get() = fieldType

    override fun toHumanClassIdentifier(className: String): String = "$className#$simpleSignature"
}